### ISSUE TITLE
[RFC] YOLO v5

### ISSUE THREAD

**Issue by oke-aditya** [Quick-AI/quickvision#48]
## 🚀 Feature

Implement YOLO v5 from `torch.hub`.
This library removes such `dataset` abstraction and aims to provides a clean modular interface to models.

Some key points to note: -
1. YOLO Achieves SOTA using Mosiac transforms as @zhiqwang mentioned.
2. We don't need to define the model/weights/backbones, we can load using torch hub.
3. We should provide an easy to use model definition, a simple `train_step`, `val_step`, `fit` API and a lightning trainer. Datasets, augmentations, transforms are not needed.

Note that none of quickvision models can achieve SOTA, limitations being torchvision's implementations and not using transforms/datasets. But they are faster, easier and flexible to train. Something which torchvision too does. 

With this context, we can start adding YOLO v5.

Depedencies: -
1. Avoid `opencv-python` at all costs. Opencv is not like PIL a library for image reading. It is huge and has lot of subdependencies. Keeping library light will enable us to use `PyTorch` Docker containers and directly infer using `torchserve`.

Evaluation mode: -
`. We don't need `.fuse()` `.eval()` such methods. We only need to load the model from `torch.hub.
 Currently, we do not have inference scripts for any models, but surely in future #2 . So right now let's focus on training.

**Comment by zhiqwang** [comment]
Hi @oke-aditya , Recently, I refactor ultralytics's yolov5 following the philosophy of torchvision's [here](https://github.com/zhiqwang/yolov5-rt-stack). Now the inference part looks similar to torchvision's retinanet. The inference part is done. And I plan to use quickvision's API (`train_step`, `val_step`, `fit` and etc.) to implement the training part.

**Comment by zhiqwang** [comment]
If we totally use `torch.hub` to implement ultralytic's in `quickvision`, it's difficult to obtain the intermediate features.

**Comment by oke-aditya** [comment]
True, but that reduces codebase significantly to maintain. I would love a custom implementation though, it would need some testing and weights etc.

**Comment by oke-aditya** [comment]
Just saw your implementation. That's great. Really nice and so complete.

**Comment by zhiqwang** [comment]
And the difference of the model weights between my implementation and ultralytics's is just some keyname, you can check the model weights script [here](https://github.com/zhiqwang/yolov5-rt-stack/blob/master/utils/updated_checkpoint.py).

**Comment by oke-aditya** [comment]
That's absoltely fine!. Just one small doubt. Will using `GeneralizedRCNNTransform` detoriate the results ?

**Comment by zhiqwang** [comment]
> Just one small doubt. Will using `GeneralizedRCNNTransform` detoriate the results ?

True, Ultralytics use the [`letterbox`](https://github.com/ultralytics/yolov5/blob/394131c2aa66637177890a41654f75ceae61a0c8/utils/datasets.py#L718) to do this function, and it reply on `opencv` [here](https://github.com/ultralytics/yolov5/blob/394131c2aa66637177890a41654f75ceae61a0c8/utils/datasets.py#L747), we should rector this funciton using torchvision to make it consistently comparing ultralytics.

**Comment by oke-aditya** [comment]
I will try once locally by using `torch.hub` without opencv dependency and try to produce a training script, which we can use.

**Comment by oke-aditya** [comment]
Though `GeneralizedRCNNTransform` worked actually fine while torchvision adopted RetinaNet, there seemed to be no issue that time. I'm not sure how super flexible it is.

**Comment by zhiqwang** [comment]
> I will try once locally by using `torch.hub` without opencv dependency and try to produce a training script, which we can use.

Hopefully to see your experiment result here.

**Comment by oke-aditya** [comment]
Sure, I will share a Colab so that we can dig into their implementation.
This will enable us to compare ours as well later !

**Comment by oke-aditya** [comment]
Just a doubt. FRCNN and RetinaNet are not differentiable in `eval` model.
E.g. 
```
model = FRCNN()
model.eval()
opt = Adam(lr=1e-3, model.parameters())
opt.step()
```

Will not work.

But it will work for Detr. Which is differentiable in both `train` and `eval` mode.

I would like to know For YOLO what are your thoughts?

**Comment by zhiqwang** [comment]
The `frcnn` and `retinanet` in `eval` mode contains `PostProcess` which shouldn't be differentiable. In other words, there are not any parameters to be learn in the module [`PostProcess`](https://github.com/pytorch/vision/blob/4521f6d152875974e317fa247a633e9ad1ea05c8/torchvision/models/detection/retinanet.py#L566).

But in `detr`, they split `PostProcess` out [here](https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/detr.py#L352), its output is same with `train` mode except for something like `BatchNormlization`.

For yolov5, we can choose either ways, No rules of thumb I think?

**Comment by oke-aditya** [comment]
No Thumb rule here.
For Detr, I seperated out the `PostProcess` as I think it is something optional and leave it to best of users.
That's what Detr too did in their paper and demo.

Best to leave it out, as I think it will be great to keep same interface in `train` and `eval` mode, this offers flexibility to users.
E.g. For CNNs and Detr, we have almost identical outputs, loss calculations in `val` mode,

Coz the idea is `training` and `validation` is really almost something similar with small changes.

While `inference` is something where we don't care about loss and do `PostProcess` etc.

**Comment by zhiqwang** [comment]
Here is a discussion of the `frcnn` output in torchvision: <https://github.com/pytorch/vision/issues/1775>

**Comment by oke-aditya** [comment]
Both `loss_dict` and `detections` is fine too, but currently torchvision models don't do that.
This is acutally good idea, as users can slowly track how the detection is improving with respect to losses.

Sometimes dataset is huge and `detectoins` are simply high which can cause memory burden to save them though.

**Comment by zhiqwang** [comment]
yep, the mechanisms in detr is more flexible, but it will influence the recovery of image shape in the PostProcess procedure. But this fixes is minor, I will try it in my yolov5-rt-track repo

**Comment by oke-aditya** [comment]
One advantage of doing `[PostProcess](https://github.com/facebookresearch/detr/blob/a54b77800eb8e64e3ad0d8237789fcbf2f8350c5/models/detr.py#L258)` is that it enables to convert from YOLO to Pascal VOC. from which we can compute `IoU` and other metrics from torchvision. 

I think if we too do PostProcess internally and are able to compute these metrics, it wold be nice.

Detr PostProcess returns

```
{'scores': s, 'labels': l, 'boxes': b}
```

If we can stick with these output formats for all models in both `train_step` and `val_step` it would be standard.
Plus this allows us to compute metrics as well which we return in both the places `train_step` and `val_step`

Sadly we can't do this for torchvision models as this is coupled with the model defintion. Hence the `train_step` differs, we don't get the detection outputs only the `loss_dict`.

So we can return 

`{"loss1" : avg_loss, "loss2": avg_loss, ... , "iou" : avg_iou, "giou" : avg_giou}` 
as a standard for `train_step` and `val_step`

**Comment by zhiqwang** [comment]
Just a doubt here, Compared to `torchvision` using **[`ImageList`](https://github.com/pytorch/vision/blob/222a599e46c5d820e8c2ce63231855ca5a30a97b/torchvision/models/detection/image_list.py#L7)**, `detr` adopts **[`NestedTensor`](https://github.com/facebookresearch/detr/blob/647917626d5017e63c1217b99537deb2dcb370d6/util/misc.py#L283)** here. One reson maybe the binding with the `boxes` reshape procedure in `PostProcess`. These difference between `ImageList` and `NestedTensor` is relatively small. If we seperate out the `PostProcess`, we should use `NestedTensor` instead?

**Comment by oke-aditya** [comment]
I think Nested Tensor Utils can be re-used from our [utils](https://github.com/Quick-AI/quickvision/blob/master/quickvision/utils/nested_tensor_utils.py).
`ImageList` is really torchvision specific, and `Nested Tensor` is probably going to be better supported and is slightly better representation.

**Comment by zhiqwang** [comment]
Hi @oke-aditya Here is a slightly abstract model graph visualization comparing `yolov5` to `retinanet`, their struct is almost same now:

![image](https://user-images.githubusercontent.com/4210061/99934857-30b03580-2d9a-11eb-83bc-f02acd12d293.png)

vs

![image](https://user-images.githubusercontent.com/4210061/99934943-7a008500-2d9a-11eb-83c0-47fffea3fbba.png)

Check for more details in [my notebooks](https://github.com/zhiqwang/yolov5-rt-stack/blob/master/notebooks/visualize-jit-models.ipynb)

**Comment by oke-aditya** [comment]
Was bit away for this weekend, I will have a look definitely.

**Comment by oke-aditya** [comment]
Just a quick update @zhiqwang . 
After release 0.1 we will continue on this. (From next week). I am super eager to have your YOLO v5 here.

**Comment by zhiqwang** [comment]
@oke-aditya Sure, and we can follow `detr` from the release `0.1`.

**Comment by zhiqwang** [comment]
Hi @oke-aditya , It seems that `0.1` will be released soon, I'm current working for make `yolov5-rt-stack` to support training now.

**Comment by oke-aditya** [comment]
Yes, release is today :smile: We will come back to this.

**Comment by zhiqwang** [comment]
Hi @oke-aditya , there are some bugs in loss computation of my `yolov5` (https://github.com/zhiqwang/yolov5-rt-stack/issues/16) When this was fixed, it should be loaded as `DETR` or `retinanet`.

**Comment by oke-aditya** [comment]
Very nice. Till then I will make layers API which will make porting YOLO easier.

**Comment by zhiqwang** [comment]
Hi @oke-aditya , The yolov5 now can be used for training, checkout https://github.com/zhiqwang/yolov5-rt-stack/pull/25 to see the details, but this may not some hidden bugs, I will add more unittest soon.

**Comment by oke-aditya** [comment]
Superb. I have been really busy this month. I will get back to working at end of month.
